parsers/cpreprocessor.c

```diff
@@ -1244,54 +1300,380 @@ static void saveIgnoreToken(const char * ignoreToken, const char * param_name)
+					
+					if(
+							(bIsVarArg && (paramLen == 3) && (strncmp(paramBegin[i],"...",3) == 0)) ||
+							((!bIsVarArg) && (paramLen == tokenLen) && (strncmp(paramBegin[i],tokenBegin,paramLen) == 0))
+						)
+					{
+						// parameter!
+						cppMacroReplacementPartInfo * rep = (cppMacroReplacementPartInfo *)eMalloc(sizeof(cppMacroReplacementPartInfo));
+						rep->parameterIndex = i;
+						rep->flags = nextParameterReplacementFlags | (bIsVarArg ? CPP_MACRO_REPLACEMENT_FLAG_VARARGS : 0);
+						rep->constant = NULL;
+						rep->next = NULL;
+
+						nextParameterReplacementFlags = 0;
+
+						CXX_DEBUG_PRINT("Parameter replacement part: %d (vararg %d)",i,bIsVarArg);
+
+						ADD_REPLACEMENT_NEW_PART(rep);
+						break;
+					}
+				}
+				
+				if(i >= iParamCount)
+				{
+					// no parameter found
+					ADD_CONSTANT_REPLACEMENT(tokenBegin,tokenLen);
+				}
+				
+				begin = c;
+				continue;
+			}
+
+			if((*c == '"') || (*c == '\''))
+			{
+				// skip string/char constant
+				char term = *c;
+				c++;
+				while(*c)
+				{
+					if(*c == '\\')
+					{
+						c++;
+						if(*c)
+							c++;
+					} else if(*c == term)
+					{
+						c++;
+						break;
+					}
+					c++;
+				}
+				continue;
+			}
+
+			if(*c == '#')
+			{
+				// check for token paste/stringification
+				if(c > begin)
+					ADD_CONSTANT_REPLACEMENT(begin,c - begin);
+
+				c++;
+				if(*c == '#')
+				{
+					// token paste
+					CXX_DEBUG_PRINT("Found token paste operator");
+					while(*c == '#')
+						c++;
+					
+					// we just skip this part and the followin spaces
+					while(*c && isspacetab(*c))
+						c++;
+					
+					if(lastReplacement && (lastReplacement->parameterIndex == -1))
+					{
+						// trim spaces from the last replacement constant!
+						const char * v = vStringValue(lastReplacement->constant);
+						const char * p = v + vStringLength(lastReplacement->constant);
+						while(p > v)
+						{
+							p--;
+							if(!isspacetab(*p))
+							{
+								p++;
+								break;
+							}
+						}
+
+						vStringTruncate(lastReplacement->constant,p-v);
+						CXX_DEBUG_PRINT("Last replacement truncated to '%s'",vStringValue(lastReplacement->constant));
+					}
+				} else {
+					// stringification
+					CXX_DEBUG_PRINT("Found stringification operator");
+					nextParameterReplacementFlags |= CPP_MACRO_REPLACEMENT_FLAG_STRINGIFY;
+				}
+				
+				begin = c;
+				continue;
+			}
+
+			c++;
+		}
+		
+		if(c > begin)
+			ADD_CONSTANT_REPLACEMENT(begin,c - begin);
+	}
+
+	hashTablePutItem(defineMacroTable,eStrndup(identifierBegin,identifierEnd - identifierBegin),info);
+	CXX_DEBUG_LEAVE();
+}
+
+static void freeMacroInfo(cppMacroInfo * info)
 {
 	if(!info)
 		return;
-	if(info->replacement)
-		vStringDelete (info->replacement);
+	if(info->replacements)
+	{
+		cppMacroReplacementPartInfo ** p = &(info->replacements);
+		while(*p)
+		{
+			if((*p)->constant)
+				vStringDelete((*p)->constant);
+			*p = (*p)->next;
+		}
+	}
```

Comment:
why not simply this?
```
for (cppMacroReplacementPartInfo *r = info->replacements; r; r = r->next)
{
    if (r->constant)
        vStringDelete(r->constant);
}
```

`info` is freed anyway, and if it's just to get `info->replacements` `NULL` at the end, it could simply be explicit instead of making "hiding" it behind the pointer indirection -- but again, unnecessary here as the location is freed right after.